projects/packages/password-checker/src/class-password-checker.php

```diff
@@ -0,0 +1,1303 @@
+	 *
+	 * @param String $string the string to be broken down.
+	 * @param String $explode_delimiter delimiter.
+	 * @return NULL|Array array of fragments, or NULL on empty string.
+	 */
+	protected function add_user_strings_to_test( $string, $explode_delimiter = ' ' ) {
+
+		// Don't check against empty strings.
+		if ( empty( $string ) ) {
+			return;
+		}
+
+		$strings = explode( $explode_delimiter, $string );
+
+		// Remove any non alpha numeric characters from the strings to check against.
+		foreach ( $strings as $key => $_string ) {
+			$strings[ $key ] = preg_replace( '/[^a-zA-Z0-9]/', '', $_string );
+		}
+
+		// Check the original too.
+		$strings[] = $string;
+
+		// Check the original minus non alpha numeric characters.
+		$strings[] = preg_replace( '/[^a-zA-Z0-9]/', '', $string );
+
+		// Remove any empty strings.
+		$strings                    = array_filter( $strings );
+		$this->user_strings_to_test = array_merge( $this->user_strings_to_test, $strings );
+	}
+
+	/**
+	 * Return a character set size that is used in the string.
+	 *
+	 * @param String $password the password.
+	 * @return Integer number of different character sets in use.
+	 */
+	protected function get_charset_size( $password ) {
+		$size = 0;
+
+		// Lowercase a-z.
+		if ( preg_match( '/[a-z]/', $password ) ) {
+			$size += 26;
+		}
+
+		// Uppercase A-Z.
+		if ( preg_match( '/[A-Z]/', substr( $password, 1, -1 ) ) ) {
+			$size += 26;
+		}
+
+		// Digits.
+		if ( preg_match( '/\d/', substr( $password, 1, -1 ) ) ) {
+			$size += 10;
+		}
+
+		// Over digits symbols.
+		if ( preg_match( '/[!|@|#|$|%|^|&|*|(|)]/', $password ) ) {
+			$size += 10;
+		}
+
+		// Other symbols.
+		if ( preg_match( '#[`|~|-|_|=|+|\[|{|\]|}|\\|\|;:\'",<\.>/\?]#', $password ) ) {
+			$size += 20;
+		}
+
+		// Spaces.
+		if ( strpos( $password, ' ' ) ) {
+			$size++;
+		}
+
+		return $size;
+	}
+
+	/**
+	 * Shorthand for getting a character index.
+	 *
+	 * @param String $char character.
+	 * @return Integer the character code.
+	 */
+	protected function get_char_index( $char ) {
+		$char = strtolower( $char[0] );
+		if ( $char < 'a' || $char > 'z' ) {
+			return 0;
+		} else {
+			return ord( $char[0] ) - ord( 'a' ) + 1;
+		}
+	}
+
+	/**
+	 * This is the password strength calculation algorithm, based on the formula H = L(logN/log2).
+	 *
+	 * H = Entropy
+	 * L = String length (the for iterator)
+	 * N = Our charset size, via get_charset_size()
+	 *
+	 * @see https://en.wikipedia.org/wiki/Password_strength#Random_passwords
+	 *
+	 * On top of the base formula, we're also multiplying the bits of entropy for every char
+	 * by 1 - (the probabily of it following the previous char)
+	 * i.e.: the probablity of U following Q is ~0.84. If our password contains this pair of characters,
+	 * the u char will only add ( 0.16^2 * charset_score ) to our total of entropy bits.
+	 *
+	 * @param String $password the password.
+	 */
+	protected function calculate_entropy_bits( $password ) {
```

Comment:
How does this compare to something like [zxcvbn](https://packagist.org/packages/bjeavons/zxcvbn-php)?

Reply:
Most likely the zxcvbn package would do a better job.
I think the idea here is to keep things simple.

Reply:
Shrug. Seems simpler to me to use an existing package than to write 839 lines of new code, plus tests.

Reply:
Keeping it simple as in this code already exists in Jetpack. Future enhancements can be done for enhancing tests.